File: mobile/sharedLogic/src/commonTest/kotlin/org/example/project/GreetingClientTest.kt
package org.example.project

import io.ktor.client.HttpClient
import io.ktor.client.engine.mock.MockEngine
import io.ktor.client.engine.mock.respond
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpStatusCode
import io.ktor.http.headersOf
import io.ktor.serialization.kotlinx.json.json
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json

class GreetingClientTest {

    @Test
    fun fetchGreetingParsesResponse() = runTest {
        val mockEngine =
            MockEngine { request ->
                assertEquals("/api/greeting", request.url.encodedPath)
                assertEquals("Android", request.url.parameters["name"])
                respond(
                    content = """{"message":"Hello, Android, from a Spring Modulith module."}""",
                    status = HttpStatusCode.OK,
                    headers = headersOf(HttpHeaders.ContentType, "application/json"),
                )
            }

        val httpClient =
            HttpClient(mockEngine) {
                install(ContentNegotiation) {
                    json(Json { ignoreUnknownKeys = true })
                }
            }

        val client = GreetingClient("http://localhost:8080", httpClient)
        assertEquals(
            "Hello, Android, from a Spring Modulith module.",
            client.fetchGreeting("Android"),
        )
    }
}

File: web/src/api/greetingClient.test.ts
import { describe, expect, it, vi } from "vitest"

import { GreetingClient } from "@/api/greetingClient"

describe("GreetingClient", () => {
  it("calls GET /api/greeting with the name query param", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify({ message: "Hello, Web, from a Spring Modulith module." }), {
        status: 200,
        headers: { "Content-Type": "application/json" },
      }),
    )

    const client = new GreetingClient("http://localhost:8080", fetchFn)
    const message = await client.fetchGreeting("Web")

    expect(message).toBe("Hello, Web, from a Spring Modulith module.")
    expect(fetchFn).toHaveBeenCalledOnce()
    const requested = String(fetchFn.mock.calls[0]?.[0])
    expect(requested).toBe("http://localhost:8080/api/greeting?name=Web")
  })

  it("throws when the response is not ok", async () => {
    const fetchFn = vi.fn().mockResolvedValue(new Response("", { status: 500 }))
    const client = new GreetingClient("http://localhost:8080", fetchFn)

    await expect(client.fetchGreeting("Web")).rejects.toThrow(
      "Greeting request failed (500)",
    )
  })
})
